### ISSUE TITLE
Problem Analyzing Functions Having the Same Names with External APIs

### ISSUE THREAD

**Issue by seviezhou** [SVF-tools/SVF#986]
I met a wired bug when using SVF, if a function has the same name and the same number of parameters with an external API, the value flow is broken for the function. For example, consider the following code (`arg.c`):

```C
#include <stdlib.h>

void *xcalloc(size_t len)
{
    void *ret;
    ret = malloc(len);
    return ret;
}

int main(int argc,char** argv)
{
    void* ptr = xcalloc(argc);
    free(ptr);
    return 0;
}
```

Compile it and check the IR with saber there will be not leaks:

```bash
clang -g -S -emit-llvm arg.c    # named arg1.ll
saber -leak arg1.ll
# not bug
```

However, when I added a second parameter to `xcalloc` function and no other operations regards to the second parameter:

```C
#include <stdlib.h>

void *xcalloc(size_t len, size_t i)
{
    void *ret;
    ret = malloc(len);
    return ret;
}

int main(int argc,char** argv)
{
    void* ptr = xcalloc(argc, argc);
    free(ptr);
    return 0;
}
```

Compile it and check the IR with saber there will be a leak:

```bash
clang -g -S -emit-llvm arg.c   # named arg2.ll
saber -leak arg2.ll
# NeverFree : memory allocation at : ({ ln: 6  cl: 11  fl: arg2.c })
```

I found that the problem lies in the value flow of function `xcalloc`, for `arg1.ll`, the value flow graph looks normal:

![arg1](https://user-images.githubusercontent.com/15180789/210775150-2d10527d-f1e4-498b-b569-c5b9edd6e055.png)

But for `arg2.ll` the value flow is broken for the returned pointer of `malloc`:

![arg2](https://user-images.githubusercontent.com/15180789/210775317-3e89ba4f-4690-47c3-9d9f-4d5882a81204.png)

The `StoreVFGNode 18` should be connected with `LoadVFGNode 16` but it does not. This bug seems to only appears when the `xcalloc` function has two parameters, less than or more than two will not trigger it.

[arg.tar.gz](https://github.com/SVF-tools/SVF/files/10351944/arg.tar.gz)

**Comment by yuleisui** [comment]
Hmm, this is a good finding. It should be related to ExtAPI.json. @shuangxiangkan can you take a look at this issue?

**Comment by yuleisui** [comment]
I suspect xcalloc with two parameters is not in the side-effect .json file but xcalloc with one parameter is in the side-effect.

**Comment by seviezhou** [comment]
Any number of parameters other than two works fine. I think there might be some code that matches functions with knowledge in `ExtAPI.json` that have the same name and the same number of parameters.

**Comment by shuangxiangkan** [comment]
In ExtAPI.json, only the side-effect summary of xcalloc with two parameters exists, but the summary of xcalloc with one parameter does not. Moreover, the attribute "overwrite_app_function" of the summary with two parameters defaults to 1，
```
"xcalloc":  {
        "return":  "void *",
        "arguments":  "(size_t, size_t)",
        "type": "EFT_ALLOC",
        "overwrite_app_function": 1,
        "AddrStmt": {
			"src": "Obj",
			"dst": "Ret"
	}
    },
```
that is, during SVF analysis, if xcalloc with two parameters is encountered, the summary in ExtAPI.json will be used instead of xcalloc in the source code . Because there is no summary of xcalloc with one parameter, the actual xcalloc code will be analyzed during SVF analysis. In order to get correct analysis results, just change the "overwrite _app_function" to 0. In this way, SVF will analyze the xcalloc in your code,
```
"xcalloc":  {
        "return":  "void *",
        "arguments":  "(size_t, size_t)",
        "type": "EFT_ALLOC",
        "overwrite_app_function": 0,
        "AddrStmt": {
			"src": "Obj",
			"dst": "Ret"
	}
    },
```

**Comment by seviezhou** [comment]
What does the summary looks like, because if there is summary for `xcalloc`, the call site of it should be marked as source. But in the example code, only `malloc` is marked as source, so there is a false alarm for memory leak.

**Comment by yuleisui** [comment]
> What does the summary looks like, because if there is summary for `xcalloc`, the call site of it should be marked as source. But in the example code, only `malloc` is marked as source, so there is a false alarm for memory leak.

xcalloc if specified in the side-effect summary file (extapi.json) as a memory allocation, then it should be marked as a souce.

**Comment by seviezhou** [comment]
The broken value flow inside `xcalloc` prevents SVF from detecting it as a wrapper function for `malloc`.

**Comment by yuleisui** [comment]
> The broken value flow inside `xcalloc` prevents SVF from detecting it as a wrapper function for `malloc`.

Have you tried to change this line to `0`?
https://github.com/SVF-tools/SVF/blob/master/svf/include/Util/ExtAPI.json#L3286

**Comment by yuleisui** [comment]
Should be this line : https://github.com/SVF-tools/SVF/blob/master/svf/include/Util/ExtAPI.json#L3396

**Comment by seviezhou** [comment]
Yes, changing it to 0 works because SVF will re-analyze `xcalloc`.

**Comment by yuleisui** [comment]
> The broken value flow inside `xcalloc` prevents SVF from detecting it as a wrapper function for `malloc`.

Hmm, this looks to be something. If we have `xcalloc` in the summary, then this `xcalloc` function body should be ignored since it will definitely be a broken value-flow and the value unable to flow out of the user-defined `xcalloc` body. We can fix this in Saber by ignoring the malloc inside a user-defined function which is also specified in the summary. Would you be able to fix this?

**Comment by seviezhou** [comment]
A possible fix is changing the following code:

https://github.com/SVF-tools/SVF/blob/06201be32965d93d708b658c77a024d6908acd1c/svf/lib/SABER/LeakChecker.cpp#L87-L95

to:

```diff
                    else
                    {
                        // exclude sources in dead functions
-                       if (cs->getCallSite()->ptrInUncalledFunction() == false)
+                       if (!cs->getCallSite()->ptrInUncalledFunction() && !isExtCall(cs->getCallSite()->getParent()->getParent()))
                        {
                            addToSources(node);
                            addSrcToCSID(node, cs);
                        }
                    }
```

**Comment by yuleisui** [comment]
Looks good to me. Could you make a pull request and see whether it can be passed?

**Comment by seviezhou** [comment]
All checks have passed.
